lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -664,6 +664,201 @@ class BitmapCanvas extends EngineCanvas with SaveStackTracking {
+  // final clipSpace -1..1 coordinates with inverted Y Axis.
+  static const _vertexShaderTriangle = '''
+      #version 300 es
+      layout (location=0) in vec4 position;
+      layout (location=1) in vec4 color;
+      uniform vec4 u_scale;
+      uniform vec4 u_shift;
+      
+      out vec4 vColor;
+      void main() {
+        gl_Position = (position * u_scale) + u_shift;
+        vColor = color;
+      }''';
+  // This fragment shader enables Int32List of colors to be passed directly
+  // to gl context buffer for rendering by decoding RGBA8888.
+  static const _fragmentShaderTriangle = '''
+      #version 300 es
+      precision highp float;
+      in vec4 vColor;
+      out vec4 fragColor;
+      void main() {
+        fragColor = vec4(vColor[2], vColor[1], vColor[0], vColor[3]);
+      }''';
+
+  /// Draws vertices on a gl context.
+  ///
+  /// If both colors and textures is specified in paint data,
+  /// for [BlendMode.source] we skip colors and use textures,
+  /// for [BlendMode.dst] we only use colors and ignore textures.
+  /// We also skip paint shader when no texture is specified.
+  ///
+  /// If no colors or textures are specified, stroke hairlines with
+  /// [Paint.color].
+  ///
+  /// If colors is specified, convert colors to premultiplied (alpha) colors
+  /// and use a SkTriColorShader to render.
+  @override
+  void drawVertices(
+      ui.Vertices vertices, ui.BlendMode blendMode, ui.PaintData paint) {
+    // TODO(flutter_web): Implement shaders for [Paint.shader] and
+    // blendMode. https://github.com/flutter/flutter/issues/40096
+    assert(paint.shader == null,
+        'Linear/Radial/SweepGradient and ImageShader not supported yet');
+    assert(blendMode == ui.BlendMode.srcOver);
+    final Int32List colors = vertices.colors;
+    final ui.VertexMode mode = vertices.mode;
+    if (colors == null) {
+      final Float32List positions = mode == ui.VertexMode.triangles
+          ? vertices.positions
+          : _convertVertexPositions(mode, vertices.positions);
+      // Draw hairline for vertices if no vertex colors are specified.
+      _drawHairline(positions, paint.color ?? ui.Color(0xFF000000));
+      return;
+    }
+
+    final html.CanvasElement glCanvas = html.CanvasElement(
+      width: _widthInBitmapPixels,
+      height: _heightInBitmapPixels,
+    );
+
+    glCanvas.style
+      ..position = 'absolute'
+      ..width = _canvas.style.width
+      ..height = _canvas.style.height;
+    glCanvas.className = 'gl-canvas';
+
+    _children.add(glCanvas);
+    rootElement.append(glCanvas);
+
+    _GlContext gl = _GlContext(glCanvas);
+
+    // Create and compile shaders.
+    Object vertexShader =
+        gl.compileShader('VERTEX_SHADER', _vertexShaderTriangle);
+    Object fragmentShader =
+        gl.compileShader('FRAGMENT_SHADER', _fragmentShaderTriangle);
+    // Create a gl program and link shaders.
+    Object program = gl.createProgram();
```

Comment:
Shader compilation per draw call is going to be really really expensive. Can these programs be cached?

Reply:
Yes very expensive. Next step is to implement blendMode which will move this to a web OffscreenCanvas and get image data instead of creating a canvas. Once it is decoupled we can cache/share program and even warm up.